File: football/Models/Standing.swift
import Foundation

// MARK: - Standings Response
struct StandingsResponse: Codable, APIErrorCheckable { // APIErrorCheckable 채택
    let get: String
    let parameters: ResponseParameters // FixtureParameters -> ResponseParameters 수정
    let errors: Any
    let results: Int
    let paging: APIPaging // FixturePaging -> APIPaging 수정
    let response: [StandingData]
    
    // 에러 필드를 딕셔너리로 변환하는 계산 속성 추가
    var errorsDict: [String: String] {
        // errors가 이미 딕셔너리인 경우
        if let errorDict = errors as? [String: String] {
            return errorDict
        }
        
        // errors가 배열인 경우
        if let errorArray = errors as? [String] {
            var dict: [String: String] = [:]
            for error in errorArray {
                // 에러 메시지에서 키-값 쌍 추출 시도
                if let colonIndex = error.firstIndex(of: ":") {
                    let key = String(error[..<colonIndex]).trimmingCharacters(in: .whitespacesAndNewlines)
                    let value = String(error[error.index(after: colonIndex)...]).trimmingCharacters(in: .whitespacesAndNewlines)
                    dict[key] = value
                } else {
                    // 콜론이 없으면 인덱스를 키로 사용
                    dict["\(dict.count)"] = error
                }
            }
            return dict
        }
        
        // 기타 타입인 경우 빈 딕셔너리 반환
        return [:]
    }
    
    // 사용자 정의 디코더 추가
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        get = try container.decode(String.self, forKey: .get)
        parameters = try container.decode(ResponseParameters.self, forKey: .parameters)
        
        // errors 필드 디코딩 (Any 타입으로 변경)
        if let errorArray = try? container.decode([String].self, forKey: .errors) {
            errors = errorArray
        } else if let errorDict = try? container.decode([String: String].self, forKey: .errors) {
            errors = errorDict
        } else {
            errors = []
        }
        
        results = try container.decode(Int.self, forKey: .results)
        paging = try container.decode(APIPaging.self, forKey: .paging)
        response = try container.decode([StandingData].self, forKey: .response)
    }
    
    // 사용자 정의 인코더 추가
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(get, forKey: .get)
        try container.encode(parameters, forKey: .parameters)
        
        // errors 필드 인코딩
        if let errorArray = errors as? [String] {
            try container.encode(errorArray, forKey: .errors)
        } else if let errorDict = errors as? [String: String] {
            try container.encode(errorDict, forKey: .errors)
        } else {
            try container.encode([] as [String], forKey: .errors)
        }
        
        try container.encode(results, forKey: .results)
        try container.encode(paging, forKey: .paging)
        try container.encode(response, forKey: .response)
    }
    
    // CodingKeys 열거형 추가
    private enum CodingKeys: String, CodingKey {
        case get, parameters, errors, results, paging, response
    }
}

// MARK: - Standing Data
struct StandingData: Codable {
    let league: StandingLeagueInfo
}

// MARK: - Standing League Info
struct StandingLeagueInfo: Codable {
    let id: Int
    let name: String
    let country: String?
    let logo: String
    let flag: String?
    let season: Int
    let standings: [[Standing]]
}

// MARK: - Standing
struct Standing: Codable, Identifiable {
    let rank: Int
    let team: StandingTeam
    let points: Int
    let goalsDiff: Int
    let group: String?
    let form: String?
    let status: String?
    let description: String?
    let all: Games
    let home: Games
    let away: Games
    let update: String
    
    var id: Int { rank }
}

// MARK: - Standing Team
struct StandingTeam: Codable {
    let id: Int
    let name: String
    let logo: String
    let country: String?
}

// MARK: - Games
struct Games: Codable {
    let played: Int
    let win: Int
    let draw: Int
    let lose: Int
    let goals: GameGoals
}

// MARK: - Game Goals
struct GameGoals: Codable {
    let goalsFor: Int
    let goalsAgainst: Int
    
    enum CodingKeys: String, CodingKey {
        case goalsFor = "for"
        case goalsAgainst = "against"
    }
}

File: Android/app/src/main/java/com/hyunwoopark/futinfo/domain/model/Standing.kt
package com.hyunwoopark.futinfo.domain.model

/**
 * 리그 순위표 도메인 모델
 */
data class LeagueStanding(
    val league: StandingLeagueInfo,
    val standings: List<List<Standing>>?
)

/**
 * 순위표 리그 정보 도메인 모델
 */
data class StandingLeagueInfo(
    val id: Int,
    val name: String,
    val country: String,
    val logo: String,
    val flag: String?,
    val season: Int
)

/**
 * 팀 순위 도메인 모델
 */
data class Standing(
    val rank: Int,
    val team: StandingTeam,
    val points: Int,
    val goalsDiff: Int,
    val group: String?,
    val form: String?,
    val status: String?,
    val description: String?,
    val all: StandingStats,
    val home: StandingStats,
    val away: StandingStats,
    val update: String
) {
    /**
     * 승률 계산
     */
    val winRate: Double
        get() = if (all.played > 0) all.win.toDouble() / all.played else 0.0
    
    /**
     * 최근 폼 상태 (W: 승, D: 무, L: 패)
     */
    val recentForm: List<String>
        get() = form?.toCharArray()?.map { it.toString() } ?: emptyList()
    
    /**
     * 홈/원정 경기력 비교
     */
    val homeAdvantage: Double
        get() {
            val homeWinRate = if (home.played > 0) home.win.toDouble() / home.played else 0.0
            val awayWinRate = if (away.played > 0) away.win.toDouble() / away.played else 0.0
            return homeWinRate - awayWinRate
        }
}

/**
 * 순위표 팀 정보 도메인 모델
 */
data class StandingTeam(
    val id: Int,
    val name: String,
    val logo: String
)

/**
 * 순위표 통계 도메인 모델
 */
data class StandingStats(
    val played: Int,
    val win: Int,
    val draw: Int,
    val lose: Int,
    val goals: StandingGoals
) {
    /**
     * 승점 계산 (승: 3점, 무: 1점, 패: 0점)
     */
    val points: Int
        get() = win * 3 + draw
    
    /**
     * 득실차 계산
     */
    val goalDifference: Int
        get() = goals.`for` - goals.against
    
    /**
     * 승률 계산
     */
    val winPercentage: Double
        get() = if (played > 0) win.toDouble() / played * 100 else 0.0
}

/**
 * 순위표 득점/실점 도메인 모델
 */
data class StandingGoals(
    val `for`: Int,
    val against: Int
) {
    /**
     * 경기당 평균 득점
     */
    fun averageGoalsFor(played: Int): Double = if (played > 0) `for`.toDouble() / played else 0.0
    
    /**
     * 경기당 평균 실점
     */
    fun averageGoalsAgainst(played: Int): Double = if (played > 0) against.toDouble() / played else 0.0
}